### PR TITLE
Created tarfs to support embedding web resources in lantern

### PR DESCRIPTION
**Pull request by oxtoacart**

### CODE REVIEW

**Comment by oxtoacart** [comment]
@fffw Would you mind taking a look at this please?

**Comment by fffw** [comment]
@oxtoacart My pleasure. Will check now.

**Comment by oxtoacart** [comment]
Thanks!

File: tarfs.go [hunk 1]
```diff
@@ -0,0 +1,251 @@
...
+func New(tarData []byte, local string) (*FileSystem, error) {
+	if local != "" {
+		_, err := os.Stat(local)
+		if err != nil {
+			if os.IsNotExist(err) {
+				log.Debugf("Local dir %v does not exist, not using\n", local)
+				local = ""
+			} else {
+				log.Errorf("Unable to stat local dir %v: %v\n", local, err)
+				os.Exit(5)
```
Comment (fffw) [inline tarfs.go]: What's the meaning of 5 here? Are there constants for exit status?

Comment (oxtoacart) [inline tarfs.go]: I'll add constants.  One sec.

Comment (oxtoacart) [inline tarfs.go]: Done

File: embed.go [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+
+	err := filepath.Walk(dir, func(path string, info os.FileInfo, err error) error {
+		if err != nil {
+			return fmt.Errorf("Unable to walk to %v: %v", path, err)
+		}
+		if info.IsDir() {
+			return nil
+		}
+		name := path
+		if strings.HasPrefix(name, dirPrefix) {
```
Comment (fffw) [inline embed.go]: `path` will always contain `dirPrefix`, no need to check here.

> The path argument contains the argument to Walk as a prefix; that is, if Walk is called with "dir", which is a directory containing the file "a", the walk function will be called with argument "dir/a".

File: tarfs_test.go [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+					_, err := io.Copy(a, f)
+					if assert.NoError(t, err, "Should be able to read from file") {
+						assert.Equal(t, expectedA, a.Bytes(), "Should have read correct data")
+					}
+				}
+			}
+		}
+	}
+}
+
```
Comment (fffw) [inline tarfs_test.go]: Though not support directory, we do and should support files under subdirectory, add test to cover it?

File: tarfs.go [hunk 4]
```diff
@@ -0,0 +1,251 @@
...
+		// Create a slice to the tar data started after the header and
+		// containing the specified size of data. We don't use tr.Read() so that
+		// we can avoid copying.
+		end := br.pos + hdr.Size
+		fs.files[hdr.Name] = tarData[br.pos:end]
+
+		// Advance to the next tar header. Note that we round up to the next
+		// multiple of 512 because tar files contain 512 byte blocks that are 0
+		// padded if the actual data doesn't align with the 512 byte boundaries.
+		err = br.AdvanceTo(int64(math.Ceil(float64(end)/512)) * 512)
```
Comment (fffw) [inline tarfs.go]: Do we need to make a test file larger than 512 bytes to cover the calculation, which is very important and subtle?

File: tarfs/main.go [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+
+func die(msg string, args ...interface{}) {
+	fmt.Fprintf(os.Stderr, msg+"\n", args...)
+	os.Exit(1)
+}
+
+func main() {
+	flag.Parse()
+	if *pkg == "" {
+		fmt.Fprintln(os.Stderr, "Please specify a pkg")
```
Comment (fffw) [inline tarfs/main.go]: `pkg` should defaults to 'main'.

File: embed.go (hunk 2 above)
Comment (oxtoacart) [inline embed.go]: Actually, if you call tarfs with the directory `.`, there is no prefix.

File: tarfs/main.go [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+	"flag"
+	"fmt"
+	"os"
+
+	"github.com/getlantern/tarfs"
+)
+
+var (
+	pkg     = flag.String("pkg", "", "The package name to use, defaults to 'main'")
+	varname = flag.String("var", "Resources", "The variable name to use, defaults to 'Data'")
```
Comment (fffw) [inline tarfs/main.go]: `var` should defaults to 'Data'

File: tarfs_test.go (hunk 3 above)
Comment (oxtoacart) [inline tarfs_test.go]: Good idea.  Done.

File: tarfs.go (hunk 4 above)
Comment (oxtoacart) [inline tarfs.go]: Good idea.  Done.

File: tarfs/main.go (hunk 5 above)
Comment (oxtoacart) [inline tarfs/main.go]: I forgot to update the flag comment.  Fixed.

File: tarfs/main.go (hunk 6 above)
Comment (oxtoacart) [inline tarfs/main.go]: I forgot to update the flag comment.  Fixed.

**Comment by fffw** [comment]
Pretty good to me now. What's the purpose of branch 'empty' here? Shouldn't we merge to master?

**Comment by oxtoacart** [comment]
@fffw Thanks for reviewing!  I did all my work on master, I created the empty branch so that I would have something against which to create a PR.  I've now merged empty to master, so we're all good!

**Pull request merged**